lib-ee/emqx_node_rebalance/src/emqx_node_rebalance.erl

```diff
@@ -0,0 +1,438 @@
+    keep_state_and_data;
+handle_event(cast, Msg, _State, _Data) ->
+    ?SLOG(warning, #{msg => "node_rebalance_unknown_cast", cast => Msg}),
+    keep_state_and_data.
+
+code_change(_Vsn, State, Data, _Extra) ->
+    {ok, State, Data}.
+
+%%--------------------------------------------------------------------
+%% internal funs
+%%--------------------------------------------------------------------
+
+enable_rebalance(#{opts := Opts} = Data) ->
+    Nodes = maps:get(nodes, Opts),
+    ConnCounts = multicall(Nodes, connection_counts, []),
+    SessCounts = multicall(Nodes, session_counts, []),
+    {_, Counts} = lists:unzip(ConnCounts),
+    Avg = avg(Counts),
+    {DonorCounts, RecipientCounts} = lists:partition(
+        fun({_Node, Count}) ->
+            Count >= Avg
+        end,
+        ConnCounts
+    ),
+    ?SLOG(warning, #{
+        msg => "node_rebalance_enabling",
+        conn_counts => ConnCounts,
+        donor_counts => DonorCounts,
+        recipient_counts => RecipientCounts
+    }),
+    {DonorNodes, _} = lists:unzip(DonorCounts),
+    {RecipientNodes, _} = lists:unzip(RecipientCounts),
+    case need_rebalance(DonorNodes, RecipientNodes, ConnCounts, SessCounts, Opts) of
+        false ->
+            {error, nothing_to_balance};
+        true ->
+            _ = multicall(DonorNodes, enable_rebalance_agent, [self()]),
+            {ok, Data#{
+                donors => DonorNodes,
+                recipients => RecipientNodes,
+                initial_conn_counts => maps:from_list(ConnCounts),
+                initial_sess_counts => maps:from_list(SessCounts)
+            }}
+    end.
+
+disable_rebalance(#{donors := DonorNodes}) ->
+    _ = multicall(DonorNodes, disable_rebalance_agent, [self()]),
+    ok.
+
+evict_conns(#{donors := DonorNodes, recipients := RecipientNodes, opts := Opts} = Data) ->
+    DonorNodeCounts = multicall(DonorNodes, connection_counts, []),
+    {_, DonorCounts} = lists:unzip(DonorNodeCounts),
+    RecipientNodeCounts = multicall(RecipientNodes, connection_counts, []),
+    {_, RecipientCounts} = lists:unzip(RecipientNodeCounts),
+
+    DonorAvg = avg(DonorCounts),
+    RecipientAvg = avg(RecipientCounts),
+    Thresholds = thresholds(conn, Opts),
+    NewData = Data#{
+        donor_conn_avg => DonorAvg,
+        recipient_conn_avg => RecipientAvg,
+        donor_conn_counts => maps:from_list(DonorNodeCounts),
+        recipient_conn_counts => maps:from_list(RecipientNodeCounts)
+    },
+    case within_thresholds(DonorAvg, RecipientAvg, Thresholds) of
+        true ->
+            ok;
+        false ->
+            ConnEvictRate = maps:get(conn_evict_rate, Opts),
+            NodesToEvict = nodes_to_evict(RecipientAvg, DonorNodeCounts),
+            ?SLOG(warning, #{
+                msg => "node_rebalance_evict_conns",
+                nodes => NodesToEvict,
+                counts => ConnEvictRate
+            }),
+            _ = multicall(NodesToEvict, evict_connections, [ConnEvictRate]),
```

Comment:
Nit: `ConnEvictRate` is actually a number of connection and not a rate, right?

Reply:
It is number of connections to evict per interval, so this is rate in some sense.